core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java

```diff
@@ -0,0 +1,275 @@
+   * @throws InvalidPathException if the alluxio path can't be resolved to a UFS mount
+   * @see {@link #fetchChildrenIfAbsent(AlluxioURI, MountTable, boolean)}
+   */
+  public Collection<UfsStatus> fetchChildrenIfAbsent(AlluxioURI path, MountTable mountTable)
+      throws IOException, InvalidPathException {
+    return fetchChildrenIfAbsent(path, mountTable, true);
+  }
+
+  /**
+   * Retrieves the child UFS statuses for a given path and stores them in the cache.
+   *
+   * This method first checks if the children have already been retrieved, and if not, then
+   * retrieves them.
+
+   * @param path the path to get the children for
+   * @param mountTable the Alluxio mount table
+   * @return the child statuses that were stored in the cache, or null if the UFS couldn't list the
+   *         statuses
+   * @throws InvalidPathException when the table can't resolve the mount for the given URI
+   */
+  @Nullable
+  private Collection<UfsStatus> getChildrenIfAbsent(AlluxioURI path, MountTable mountTable)
+      throws InvalidPathException {
+    Collection<UfsStatus> children = getChildren(path);
+    if (children != null) {
+      return children;
+    }
+    MountTable.Resolution resolution = mountTable.resolve(path);
+    AlluxioURI ufsUri = resolution.getUri();
+    try (CloseableResource<UnderFileSystem> ufsResource = resolution.acquireUfsResource()) {
+      UnderFileSystem ufs = ufsResource.get();
+      UfsStatus[] statuses = ufs.listStatus(ufsUri.toString());
+      if (statuses == null) {
+        return null;
+      }
+      children = Arrays.asList(statuses);
+      addChildren(path, children);
+    } catch (IllegalArgumentException | IOException e) {
+      LOG.debug("Failed to add status to cache", e);
+    }
+    return children;
+  }
+
+  /**
+   * Get the child {@link UfsStatus}es from a given {@link AlluxioURI}.
+   *
+   * @param path the path the retrieve
+   * @return The corresponding {@link UfsStatus} or {@code null} if there is none stored
+   */
+  @Nullable
+  public Collection<UfsStatus> getChildren(AlluxioURI path) {
+    return mChildren.get(path);
+  }
+
+  /**
+   * Submit a request to asynchronously fetch the statuses corresponding to a given directory.
+   *
+   * Retrieve any fetched statuses by calling {@link #fetchChildrenIfAbsent(AlluxioURI, MountTable)}
+   * with the same Alluxio path.
+   *
+   * If no {@link ExecutorService} was provided to this object before instantiation, this method is
+   * a no-op.
+   *
+   * @param path the path to prefetch
+   * @param mountTable the Alluxio mount table
+   */
+  public void prefetchChildren(AlluxioURI path, MountTable mountTable) {
+    if (mPrefetchExecutor == null) {
+      return;
+    }
+    try {
+      Future<Collection<UfsStatus>> fute =
+          mPrefetchExecutor.submit(() -> getChildrenIfAbsent(path, mountTable));
+      Future<Collection<UfsStatus>> prev = mActivePrefetchJobs.put(path, fute);
+      if (prev != null) {
+        prev.cancel(true);
```

Comment:
Why cancel the old one? Could this cause an endless cycle of constantly cancelled tasks?

Reply:
How could it cause an endless cycle?

It only returns the previously submitted job for the given path. We have two options if there is already a prefetch job running on that path.

1. (current approach) Cancel the old one, submit the new one
2. Ignore the current request and rely on the future for the previously submitted future.

This method shouldn't really occur in the current code, but in the case it does happen I wanted to define the behavior. If the task is submitted later, I make the assumption that the user is submitting it because of some possible updated data in the UFS. The current prefetch job may already be executing or may be finished. I cancel it anyway in case it is running or about to run, and then submit the next one so whoever goes looking for the children later on will have a likely more up-to-date version of the UfsStatuses.

I don't see how this could cause an endless cycle?

Reply:
> Well, if we keep calling prefetchChildren, it will never terminate
I don't understand what is "never terminating"?

calling this method only schedules the job and immediately returns. It may cancel a previous one if it exists. This method will never block.

we only ever call `prefetchChildren` once in the code, and it doesn't depend on whether or not there is a job in the cache for that path. For the code paths that use that job to retrieve the child statuses, they also don't have a strong dependency on the prefetch job existing or not. If the job they wait for gets cancelled (or doesn't exist) it will just make the call to the UFS serially.

Reply:
In `fetchChildrenIfAbsent`, it is supposed to wait for the prefetch job. If the prefetch job is always getting canceled and re-created, I'd imagine the prefetch job would never finish, and that i where the "cycle" would happen.

But it sounds like you are saying `prefetchChildren` is only called once per path per instance of this class?

Reply:
Yes, it should only ever happen once per path in the current code